src/VoiceProtocol.h

```diff
@@ -0,0 +1,47 @@
+// Copyright 2005-2020 The Mumble Developers. All rights reserved.
+// Use of this source code is governed by a BSD-style license
+// that can be found in the LICENSE file at the root of the
+// Mumble source tree or at <https://www.mumble.info/LICENSE>.
+
+#ifndef MUMBLE_VOICEPROTOCOLTYPE_H
+#define MUMBLE_VOICEPROTOCOLTYPE_H
+
+#include "crypto/CipherType.h"
+#include "crypto/CryptStateFactory.h"
+
+#include <exception>
+#include <memory>
+#include <string>
+
+enum class VoiceTransportType { INVALID = 0, UDP = 1 };
+
+class VoiceProtocolException : public std::exception {};
+
+class VoiceProtocol {
+protected:
+	bool m_valid;
+	std::string m_protocolName;
+	explicit VoiceProtocol(const std::string &s);
+
+public:
+	VoiceTransportType m_transport;
+	virtual bool operator==(const VoiceProtocol &rhs) const;
+	virtual std::string toString() const;
+	static std::shared_ptr< VoiceProtocol > fromString(const std::string &s);
+	bool isValid() const;
+	VoiceProtocol();
+	virtual ~VoiceProtocol(){};
+};
+
+
+class UDPVoiceProtocol : public VoiceProtocol {
+private:
+	explicit UDPVoiceProtocol(const std::string &s);
+	friend VoiceProtocol;
```

Comment:
Why is this needed? The class already inherits `VoiceProtocol`, so one could assume making them friends is unnecessary? :thinking:

Reply:
I think the problem is I want to call UDPVoiceProtocol's function in VoiceProtocol... See
https://github.com/mumble-voip/mumble/blob/d20d3e91ce08b7a5b11605c8a1aabcb083aadb5d/src/VoiceProtocol.cpp#L21-L24

I know it doesn't feel right. And actually my solution to similar problems (see the crypto part) was to create another factory object to do the conversion... Maybe I should do it here as well?

Reply:
Ah so you are calling a private constructor. Okay then that makes sense then :bulb: 

Note however that you are misusing `make_shared` (slightly). It should look like this:
```cpp
return std::make_shared< UDPVoiceProtocol >(s);
```
The way you wrote it, it first creates an instance of the class and then copies (or moves) that to another instance which is then contained in the shared pointer. With my version you are only creating a single object and don't copy/move a second one around :)